tests/bluetooth/bsim/host/adv/encrypted/src/central.c

```diff
@@ -0,0 +1,90 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "common.h"
+
+#define EXPECTED_DATA_SIZE 6
+static uint8_t expected_data[] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};
+BUILD_ASSERT(sizeof(expected_data) == EXPECTED_DATA_SIZE);
+
+static uint8_t session_key[] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
+				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
+BUILD_ASSERT(sizeof(session_key) == BT_EAD_KEY_SIZE);
+static uint8_t iv[] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
+BUILD_ASSERT(sizeof(iv) == BT_EAD_IV_SIZE);
+
+static bt_addr_le_t prev_addr;
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	int err;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	LOG_DBG("Device found: %s (RSSI %d)", addr_str, rssi);
+
+	if (!bt_addr_le_eq(addr, &prev_addr)) {
+		uint8_t *encrypted_data =
+			&ad->data[2]; /* the ad data are after the length and the type */
```

Comment:
How do we tell if the data is encrypted, before attempting to decrypt it?

Reply:
If we receive an advertising structure with the Ad Type `Encrypted Advertising Data`, then the Ad Data of this structure are encrypted.

Reply:
Ah, nice. We should check the type before decrypting then, right?

Reply:
What I suggest is that we take the `ad` here and forward it to `bt_data_parse`. Then in the `bt_data_parse` callback, we check the AD type to see if it is encrypted, and only then decrypt it. 

That would be more similar to how an application would/should use this, and would also avoid some of the length handling here.